### PR TITLE
describe image pull secrets

### PR DESCRIPTION
**Pull request by deads2k**
Describes how to create and use pull secrets with service accounts.

@liggitt @ncdc ptal

### CODE REVIEW

**Comment by adellape** [comment]
@deads2k Content LGTM. How does this contrast with the following Dev Guide section?

http://docs.openshift.org/latest/dev_guide/builds.html#using-docker-credentials-for-pushing-and-pulling-images

Is the above (or at least portions of it) deprecated? Are there use cases for both methods, or should end-users always use service accounts?

File: admin_guide/image_pull_secrets.adoc [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+== Overview
+Docker registries can be secured to prevent unauthorized parties from accessing certain images.  If you are using the OpenShift docker registry and are pulling images for pods located in the same project, then your service account should already have permissions and no additional action should be required.  If this is not the case, then additional configuration steps are required.
+
+=== OpenShift Docker Registry Authentication and Authorization
+The OpenShift Docker Registry authenticates using the same tokens as the OpenShift API.  To perform a `docker login` against the OpenShift Docker registry, you can choose any username and email, but the password must be a valid OpenShift token.  In order to pull an image, the authenticated user must have `get` rights on the requested `imagestreams/layer`.  In order to push an image, the authenticated user must have `update` rights on the requested `imagestreams/layer`.
+
+By default, all service accounts in a project have rights to pull any image in the same project and the builder service account has rights to push any image in the same project.
+
+=== Allowing Pods in ProjectA to reference images from ProjectB
```
Comment (liggitt) [inline admin_guide/image_pull_secrets.adoc]: globally, use valid project names in examples (lowercase, alphanumeric with hyphen separators)

File: admin_guide/image_pull_secrets.adoc [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+In order to allow pods in ProjectA to reference images in ProjectB, a service account in ProjectA should be bound to the `system:image-puller` role in ProjectB (`oc policy add-role-to-user serviceaccount:ProjectA:default --namespace=ProjectB`).  After adding that role, the pods in ProjectA that reference the default service account will be able to pull images from Project B.
```
Comment (liggitt) [inline admin_guide/image_pull_secrets.adoc]: command should be
```
oc policy add-role-to-user system:image-puller serviceaccount:project-a:default --namespace=project-b
```

Comment (liggitt) [inline admin_guide/image_pull_secrets.adoc]: to allow any service account in project-a access:
```
oc policy add-role-to-group system:image-puller system:serviceaccounts:project-a --namespace=project-b
```

File: admin_guide/image_pull_secrets.adoc [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+Once you've created your secret, you need to add that secret to your service account.
+
+If you are going to use your secret for pulling images for pods, then you can add the secret to your service account by running
+```
+oc secrets add serviceaccount/default secrets/my-pull-secret --for=pull
+```
+
+If you are going to use your secret for push and pulling build images, then the secret needs to be mountable inside of a pod.  You can do this by running
+```
+oc secrets add serviceaccount/default secrets/my-pull-secret
```
Comment (liggitt) [inline admin_guide/image_pull_secrets.adoc]: should this be the `builder` service account, if the example is for running builds?

File: admin_guide/image_pull_secrets.adoc [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+If you don't already have a .dockercfg file, you can create a secret by running
+```
+oc secrets new-dockercfg my-pull-secret --docker-server=DOCKER_REGISTRY_SERVER --docker-username=DOCKER_USER --docker-password=DOCKER_PASSWORD --docker-email=DOCKER_EMAIL
+```
+
+Once you've created your secret, you need to add that secret to your service account.
+
+If you are going to use your secret for pulling images for pods, then you can add the secret to your service account by running
+```
+oc secrets add serviceaccount/default secrets/my-pull-secret --for=pull
```
Comment (liggitt) [inline admin_guide/image_pull_secrets.adoc]: the name of the service account in this example should match the name of the service account the pod will use ("default" is the default service account)

File: admin_guide/image_pull_secrets.adoc [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+Docker registries can be secured to prevent unauthorized parties from accessing certain images.  If you are using the OpenShift docker registry and are pulling images for pods located in the same project, then your service account should already have permissions and no additional action should be required.  If this is not the case, then additional configuration steps are required.
+
+=== OpenShift Docker Registry Authentication and Authorization
+The OpenShift Docker Registry authenticates using the same tokens as the OpenShift API.  To perform a `docker login` against the OpenShift Docker registry, you can choose any username and email, but the password must be a valid OpenShift token.  In order to pull an image, the authenticated user must have `get` rights on the requested `imagestreams/layer`.  In order to push an image, the authenticated user must have `update` rights on the requested `imagestreams/layer`.
```
Comment (liggitt) [inline admin_guide/image_pull_secrets.adoc]: `imagestreams/layers`

**Comment by deads2k** [comment]
> Is the above (or at least portions of it) deprecated? Are there use cases for both methods, or should end-users always use service accounts?

The two documents are describing slightly different things.  This doc is about how to make use of pull secrets to retrieve container images.  The builds document is specifically about how to use pull secrets for builds.  I did update it to use the current commands.

File: dev_guide/builds.adoc [hunk 6]
```diff
@@ -492,13 +492,21 @@ following command:
...
+'dockerhub' and creates the object.
+
+Once you have created your Secret, it must be added to the builder service account
+in order to be used.
+
+===
+----
+$ oc secrets add serviceaccount/builder secrets/dockerhub
+----
+===
```
Comment (adellape) [inline dev_guide/builds.adoc]: Should be `====`s.

**Comment by adellape** [comment]
@deads2k Thanks for the clarification. One minor issue inline.

**Comment by adellape** [comment]
@deads2k Is there any reason the new `image_pull_secrets` topic couldn't be in the Dev Guide instead? If not, I can mess w/ that in a follow-up w/ other minor edits.

**Comment by adellape** [comment]
Merging as-is after confirming with @liggitt and will follow-up with edits (e.g., moving `image_pull_secrets` to Dev Guide).

**Pull request merged**